staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go

```diff
@@ -1375,6 +1383,30 @@ func mergeMap(original, patch map[string]interface{}, schema LookupPatchMeta, me
 	return original, nil
 }
 
+// discardNullValuesFromPatch discards all null values from patch.
+// It traverses for all slices and map types to discard all nulls.
+func discardNullValuesFromPatch(patchV interface{}) interface{} {
+	if patchV == nil {
+		return nil
+	}
+
+	patchT := reflect.TypeOf(patchV)
+	switch patchT.Kind() {
+	case reflect.Map:
+		filteredPatch := make(map[string]interface{})
+		for k, v := range patchV.(map[string]interface{}) {
+			if f := discardNullValuesFromPatch(v); f != nil {
+				filteredPatch[k] = f
+			}
+		}
+
+		return filteredPatch
+
+	default:
+		return patchV
+	}
```

Comment:
A few observations:
1. I'm not sure what the reflection is getting us that a direct type cast attempt wouldn't do
2. Kind() == Map doesn't guarantee it is a `map[string]interface{}`, right?
3. The only calls here are from a function that says it can modify the patch... should we delete null elements from the patch instead of making a copy of each level?
4. The godoc says it recurses into slices, but I don't see that happening... should it be? if so, we need a test for that as well

I wonder if this could be simplified to something like:
```go
switch patchV := patchV.(type) {
case map[string]interface{}:
  for k, v := range patchV {
    if v == nil {
      delete(patchV, k)
    } else {
      discardNullValuesFromPatch(v)
    }
  }

case []interface{}:
  for _, v := range patchV {
    discardNullValuesFromPatch(v)
  }

}

return patchV
```

Reply:
> A few observations:
> 
> 1. I'm not sure what the reflection is getting us that a direct type cast attempt wouldn't do
> 2. Kind() == Map doesn't guarantee it is a `map[string]interface{}`, right?

I'm definitely in favor of using type casting instead reflection. But as far as I can see, there are lots of places using reflection in this package, and I thought that reflection is being used due to a **_reason_**.
> 3. The only calls here are from a function that says it can modify the patch... should we delete null elements from the patch instead of making a copy of each level?

Good idea
> 4. The godoc says it recurses into slices, but I don't see that happening... should it be? if so, we need a test for that as well
> 

First attempt tried to also iterate over slices but then I removed it. Maybe we should put it back. What do you think @apelisse ?
> I wonder if this could be simplified to something like:
> 
> ```go
> switch patchV := patchV.(type) {
> case map[string]interface{}:
>   for k, v := range patchV {
>     if v == nil {
>       delete(patchV, k)
>     } else {
>       discardNullValuesFromPatch(v)
>     }
>   }
> 
> case []interface{}:
>   for _, v := range patchV {
>     discardNullValuesFromPatch(v)
>   }
> 
> }
> 
> return patchV
> ```

That function ^ looks good to me.